File: services/lab-auth/internal/identity/identity.go
package identity

import "strings"

// Rules mirrored from src/boot-identity.ts (CONTRACT.md §2). The two
// implementations are held in sync by the shared fixture at
// scripts/auth/fixtures/identity-cases.json (see parity_test.go).

const (
	UsernameMin = 3
	UsernameMax = 24
	PasswordMin = 15
	PasswordMax = 128
)

var reservedUsernameKeys = []string{"guest"}

// Key is the ASCII lower-case uniqueness key. Never displayed.
func Key(username string) string { return strings.ToLower(username) }

// Label is the upper-case display value for "ID CONFIRMED : <LABEL>".
func Label(username string) string { return strings.ToUpper(username) }

// CodePoints counts Unicode code points, matching [...value].length in TS.
func CodePoints(value string) int { return len([]rune(value)) }

type UsernameError string

const (
	UsernameTooShort UsernameError = "too-short"
	UsernameTooLong  UsernameError = "too-long"
	UsernameCharset  UsernameError = "charset"
	UsernameReserved UsernameError = "reserved"
)

func isUsernameRune(r rune) bool {
	switch {
	case r >= 'A' && r <= 'Z':
		return true
	case r >= 'a' && r <= 'z':
		return true
	case r >= '0' && r <= '9':
		return true
	case r == '.' || r == '_' || r == '-':
		return true
	default:
		return false
	}
}

// ValidateUsername returns the uniqueness key, or an error string matching the
// TS UsernameError union.
func ValidateUsername(value string) (key string, err UsernameError) {
	length := CodePoints(value)
	if length < UsernameMin {
		return "", UsernameTooShort
	}
	if length > UsernameMax {
		return "", UsernameTooLong
	}
	for _, r := range value {
		if !isUsernameRune(r) {
			return "", UsernameCharset
		}
	}
	key = Key(value)
	for _, reserved := range reservedUsernameKeys {
		if key == reserved {
			return "", UsernameReserved
		}
	}
	return key, ""
}

type PasswordError string

const (
	PasswordTooShort PasswordError = "too-short"
	PasswordTooLong  PasswordError = "too-long"
)

// ValidatePassword counts code points; it never trims or normalises.
func ValidatePassword(value string) PasswordError {
	length := CodePoints(value)
	if length < PasswordMin {
		return PasswordTooShort
	}
	if length > PasswordMax {
		return PasswordTooLong
	}
	return ""
}

File: shared/auth/identity.ts
// Account rules shared by every surface (blog pages and the /lab/ archive).
//
// This module is deliberately pure: no DOM, no network, no storage. It defines
// the identity data model and the username/password rules that the Go service
// mirrors in internal/identity (see services/lab-auth/internal/identity). The
// API wire contract lives in services/lab-auth/openapi.yaml.
//
// It is imported by:
//   - apps/blog (the static account page and the header account control)
//   - src/features/auth (the boot identity gate and the 3D archive)
//   - scripts/auth fixtures, which load it through scripts/auth/load-ts.mjs
// so it must stay import-free at runtime.

export const USERNAME_MIN = 3;
export const USERNAME_MAX = 24;
export const USERNAME_PATTERN = /^[A-Za-z0-9._-]+$/;
export const RESERVED_USERNAME_KEYS: readonly string[] = ["guest"];
export const PASSWORD_MIN = 15;
export const PASSWORD_MAX = 128;

export type IdentityKind = "none" | "guest" | "registered";

export interface NoIdentity {
  readonly kind: "none";
}

export interface GuestIdentity {
  readonly kind: "guest";
  readonly label: "GUEST";
}

export interface RegisteredIdentity {
  readonly kind: "registered";
  readonly userId: string;
  readonly username: string;
  readonly label: string;
}

export type BootIdentity = NoIdentity | GuestIdentity | RegisteredIdentity;

// An identity the user has actually chosen (never the "none" placeholder).
export type ChosenIdentity = GuestIdentity | RegisteredIdentity;

export type UsernameError = "too-short" | "too-long" | "charset" | "reserved";
export type PasswordError = "too-short" | "too-long";

export interface ValidUsername {
  readonly ok: true;
  readonly key: string;
  readonly username: string;
  readonly label: string;
}

export type UsernameResult =
  ValidUsername | { readonly ok: false; readonly error: UsernameError };
export type PasswordResult =
  { readonly ok: true } | { readonly ok: false; readonly error: PasswordError };

export const NO_IDENTITY: NoIdentity = { kind: "none" };

/** Uniqueness key: ASCII lower-case. Never used for display. */
export function usernameKey(username: string): string {
  return username.toLowerCase();
}

/** Display label: upper-case, matching `ID CONFIRMED : <LABEL>`. */
export function usernameLabel(username: string): string {
  return username.toUpperCase();
}

export function usernameCodePoints(value: string): number {
  return [...value].length;
}

export function validateUsername(value: string): UsernameResult {
  const length = usernameCodePoints(value);
  if (length < USERNAME_MIN) return { ok: false, error: "too-short" };
  if (length > USERNAME_MAX) return { ok: false, error: "too-long" };
  if (!USERNAME_PATTERN.test(value)) return { ok: false, error: "charset" };
  const key = usernameKey(value);
  if (RESERVED_USERNAME_KEYS.includes(key))
    return { ok: false, error: "reserved" };
  return { ok: true, key, username: value, label: usernameLabel(value) };
}

/** Passwords are never trimmed or normalised; length is Unicode code points. */
export function passwordCodePoints(value: string): number {
  return [...value].length;
}

export function validatePassword(value: string): PasswordResult {
  const length = passwordCodePoints(value);
  if (length < PASSWORD_MIN) return { ok: false, error: "too-short" };
  if (length > PASSWORD_MAX) return { ok: false, error: "too-long" };
  return { ok: true };
}

export function guestIdentity(): GuestIdentity {
  return { kind: "guest", label: "GUEST" };
}

/**
 * Build a registered identity from a server response. The label is derived
 * from the server-returned username, never from raw form input.
 */
export function registeredIdentity(
  userId: string,
  username: string,
): RegisteredIdentity {
  return {
    kind: "registered",
    userId,
    username,
    label: usernameLabel(username),
  };
}

export function sameIdentity(a: BootIdentity, b: BootIdentity): boolean {
  if (a.kind !== b.kind) return false;
  if (a.kind === "registered" && b.kind === "registered")
    return a.userId === b.userId;
  return true;
}
